packages/website/blog/2024-09-30-typed-linting.md

```diff
@@ -0,0 +1,247 @@
+
+```ts
+const { getItem } = localStorage;
+//      ~~~~~~~
+// Avoid referencing unbound methods which may cause unintentional scoping of `this`.
+// eslint(@typescript-eslint/unbound-method)
+```
+
+That's useful because calls to `getItem` that aren't bound to `localStorage` cause an exception at runtime:
+
+```ts
+getItem('...');
+// Uncaught TypeError: Illegal invocation
+```
+
+Without type information to indicate the types of `localStorage` and its `getItem` property, there would have been no way to determine that the `const { getItem }` access was unsafe.
+
+### Async Race Conditions
+
+Even if your code is 100% typed, has no `any`s, and doesn't misuse scopes, it's still possible to have bugs that can only easily be detected by typed linting.
+Asynchronous code with `Promise`s in particular can introduce subtle issues that are completely type-safe.
+
+Suppose your code is meant to run an asynchronous `readFromCache` function before reading from the file system:
+
+```ts
+import { fs } from 'node:fs/promises';
+import { readFromCache } from './caching';
+
+const filePath = './data.json';
+
+readFromCache(filePath);)
+
+await fs.rm(filePath);
+```
+
+Do you see the potential bug?
+
+If `readFromCache` is asynchronous (returns a `Promise`), then calling it and not awaiting its returned Promise could lead to race conditions in code.
+Its asynchronous or delayed logic might not get to reading from the `filePath` before `fs.rm(filePath)` runs.
+
+This is commonly referred to as a _"floating"_ Promise: one that is created but not appropriately handled.
+The [`@typescript-eslint/no-floating-promises`](/rules/no-floating-promises) lint rule would report on that floating Promise:
+
+```ts
+readFromCache(filePath);
+// Promises must be awaited, end with a call to .catch, end with a call to .then
+// with a rejection handler or be explicitly marked as ignored with the `void` operator.
+// eslint(@typescript-eslint/no-floating-promises
+```
+
+...and can give an editor suggestion to add a missing `await`:
+
+```diff
+- readFromCache(filePath);
++ await readFromCache(filePath);
+```
+
+Determining whether code is creating a floating Promise is only possible when the types of code are known.
+Otherwise, lint rules would have no way of knowing which imports from other files could potentially create a Promise that needs to be handled.
+
+### Custom Rules
+
+Typed linting isn't restricted to just typescript-eslint rules.
+It can be used in community ESLint plugins, as well as custom rules specific to your project.
+
+One common example used by teams is to codemod from a deprecated API to its replacement.
+Typed linting is often necessary to determine which pieces of code call to the old API.
+
+As an example, consider the following `fetch()` network call.
+A typed lint rule could determine that `endpoint` is a string referring to an old endpoint:
```

Comment:
of all things I would think that this is something that you would use find-and-replace for 🤔

Reply:
Heh yeah this isn't my favorite example I've come up with. I couldn't come up with one that's quick to explain and is clearly you'd need type info for.

I was thinking `endpoint` might be a common name, or otherwise used as an alias. Like, you could read from the _literal_ type of it, if that's possible...

🤷 it's not great. Will try to think of a better one.

Reply:
All right, switched it to changing a `fetch()` call's `data`. I would personally use a wrapper but I've seen teams opt to use this before instead.